rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2215 @@
+
+### 📖 Safety
+A sound transmutation is ***safe*** if *using* the transmuted value cannot violate memory safety.
+
+### 📖 Stability
+A safe transmutation is ***stable*** if the authors of the source type and destination types have indicated that the layouts of those types is part of their libraries' stability guarantees.
+
+## Concepts in Depth
+
+***Disclaimer:** While the high-level definitions of transmutation soundness, safety and stability are a core component of this RFC, the detailed rules and examples in this section are **not**. We expect that the initial implementation of `TransmuteFrom` may initially be considerably less sophisticated than the examples in this section (and thus forbid valid transmutations). Nonetheless, this section explores nuanced cases of transmutation soundness and safety to demonstrate that the APIs we propose can grow to handle that nuance.*
+
+
+### 📖 When is a transmutation sound?
+[sound transmutation]: #When-is-a-transmutation-sound?
+A transmutation is ***sound*** if the mere act of transmuting a value from one type to another is not undefined behavior.
+
+#### Well-Defined Representation
+[`u8`]: core::u8
+[`f32`]: core::f32
+
+Transmutation is *always unsound* if it occurs between types with unspecified representations.
+
+Most of Rust's primitive types have specified representations. That is, the precise layout characteristics of [`u8`], [`f32`] is a documented and guaranteed aspect of those types.
+
+In contrast, most `struct` and `enum` types defined without an explicit `#[repr(C)]` attribute do ***not*** have well-specified layout characteristics.
+
+To ensure that types you've define are soundly transmutable, you almost always (with very few exceptions) must mark them with the `#[repr(C)]` attribute.
+
+#### Requirements on Owned Values
+[transmute-owned]: #transmuting-owned-values
+
+Transmutations involving owned values must adhere to two rules to be sound. They must:
+ * [preserve or broaden the bit validity][owned-validity], and
+ * [preserve or shrink the size][owned-size].
+
+##### Preserve or Broaden Bit Validity
+[owned-validity]: #Preserve-or-Broaden-Bit-Validity
+[`NonZeroU8`]: https://doc.rust-lang.org/beta/core/num/struct.NonZeroU8.html
+
+The bits of any valid instance of the source type must be a bit-valid instance of the destination type.
+
+For example, we are permitted to transmute a `Bool` into a [`u8`]:
+```rust
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(u8)]
+enum Bool {
+    True = 1,
+    False = 0,
+}
+
+let _ : u8 = Bool::True.transmute_into();
+let _ : u8 = Bool::False.transmute_into();
+```
+<sup>
+(Note: <code>#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]</code> annotation connotes that <i>all</i> aspects of <code>Bool</code>'s layout are part of its <a href="#When-is-a-transmutation-stable?">library stability guarantee</a>.)
```

Comment:
Broken link: #When-is-a-transmutation-stable?

Reply:
```suggestion
(Note: <code>#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]</code> annotation connotes that <i>all</i> aspects of <code>Bool</code>'s layout are part of its <a href="#-when-is-a-transmutation-stable">library stability guarantee</a>.)
```